src/main/java/com/mraof/minestuck/api/alchemy/GristSet.java

```diff
@@ -0,0 +1,136 @@
+ * There's also {@link NonNegativeGristSet}, which is a mutable grist set that throws an exception if any amount goes below zero.
+ */
+public interface GristSet
+{
+	String GRIST_COMMA = "grist.comma";
+	
+	default long getGrist(GristType type)
+	{
+		return this.asMap().getOrDefault(type, 0L);
+	}
+	
+	default boolean hasType(GristType type)
+	{
+		return this.asMap().containsKey(type);
+	}
+	
+	default boolean isEmpty()
+	{
+		return this.asMap().values().stream().allMatch(amount -> amount == 0);
+	}
+	
+	default boolean equalContent(GristSet other)
+	{
+		for(GristType type : GristTypes.values())
+			if(this.getGrist(type) != other.getGrist(type))
+				return false;
+		return true;
+	}
+	
+	/**
+	 * @return a value estimate for this grist set
+	 */
+	default double getValue()
+	{
+		double sum = 0;
+		for(GristAmount amount : asAmounts())
+			sum += amount.getValue();
+		return sum;
+	}
+	
+	/**
+	 * @return an unmodifiable map with all grist types and amounts contained in this grist set.
+	 */
+	Map<GristType, Long> asMap();
+	
+	ImmutableGristSet asImmutable();
+	
+	default MutableGristSet mutableCopy()
+	{
+		return new DefaultMutableGristSet(this);
+	}
+	
+	default List<GristAmount> asAmounts()
+	{
+		return this.asMap().entrySet().stream().map(entry -> new GristAmount(entry.getKey(), entry.getValue())).toList();
+	}
+	
+	default Component asTextComponent()
+	{
+		Component component = null;
+		for(GristAmount grist : asAmounts())
+		{
+			if(component == null)
+				component = grist.asTextComponent();
+			else component = Component.translatable(GRIST_COMMA, component, grist.asTextComponent());
+		}
+		if(component != null)
+			return component;
+		else return Component.empty();
+	}
+	
+	ImmutableGristSet EMPTY = Collections::emptyMap;
+	
+	static ImmutableGristSet of(GristType type, long amount)
+	{
+		return new GristAmount(type, amount);
+	}
+	
+	static ImmutableGristSet of(Supplier<GristType> type, long amount)
+	{
+		return new GristAmount(type, amount);
+	}
+	
+	static ImmutableGristSet of(Supplier<GristType> type1, long amount1, Supplier<GristType> type2, long amount2)
+	{
+		return new DefaultImmutableGristSet(ImmutableMap.<GristType, Long>builder().put(type1.get(), amount1).put(type2.get(), amount2));
+	}
+	
+	static ImmutableGristSet of(Supplier<GristType> type1, long amount1, Supplier<GristType> type2, long amount2, Supplier<GristType> type3, long amount3)
+	{
+		return new DefaultImmutableGristSet(ImmutableMap.<GristType, Long>builder().put(type1.get(), amount1).put(type2.get(), amount2).put(type3.get(), amount3));
+	}
+	
+	static ImmutableGristSet of(Supplier<GristType> type1, long amount1, Supplier<GristType> type2, long amount2, Supplier<GristType> type3, long amount3, Supplier<GristType> type4, long amount4)
+	{
+		return new DefaultImmutableGristSet(ImmutableMap.<GristType, Long>builder().put(type1.get(), amount1).put(type2.get(), amount2).put(type3.get(), amount3).put(type4.get(), amount4));
+	}
```

Comment:
Would it make sense to create an object and pass any number of grist type+amount together using a vararg

Reply:
varargs wouldn't work here because we need both multiple types and amounts

Reply:
Perhaps im missing something, but what I am suggesting is squishing a grist type + amount in an object so that the object can be passed in

Reply:
I guess `GristAmount`s would fit that role?

Reply:
Perhaps so.
One of the longest usages of this function is:
```java
GristSet.of(GristTypes.BUILD, 1, GristTypes.COBALT, 1, GristTypes.CHALK, 1, GristTypes.AMBER, 1)
```
Using a vararg with grist amounts as is, we would instead write this as:
```java
GristSet.of(new GristAmount(GristTypes.BUILD.get(), 1), new GristAmount(GristTypes.COBALT.get(), 1), new GristAmount(GristTypes.CHALK.get(), 1), new GristAmount(GristTypes.AMBER.get(), 1))
```
By overloading the constructor, we could reduce this to:
```java
GristSet.of(new GristAmount(GristTypes.BUILD, 1), new GristAmount(GristTypes.COBALT, 1), new GristAmount(GristTypes.CHALK, 1), new GristAmount(GristTypes.AMBER, 1))
```
If we define a function `public static GristAmount amount(long)` for grist types, we could instead write:
```java
GristSet.of(GristTypes.BUILD.get().amount(1), GristTypes.COBALT.get().amount(1), GristTypes.CHALK.get().amount(1), GristTypes.AMBER.get().amount(1))
```
Another alternative is to define a builder object:
```java
GristSet.Builder.new().with(GristTypes.BUILD, 1).with(GristTypes.COBALT, 1).with(GristTypes.CHALK, 1).with(GristTypes.AMBER, 1).build()
```

My current thought is to keep the current functions for fewer arguments (because of how compact those functions are), and add a builder for grist sets with a larger number of grist types.
What are your thoughts?

Reply:
Or actually, as I just sent that, I changed my mind and like the `amount()` functions a bit better because they group together the grist type and its corresponding amount more clearly than the current thing.

Reply:
I feel like the `amount()` functions would be both easier to implement and more readable than the builder. Plus, they could be used in other places where GristAmounts would be needed.